pkg/ccl/backupccl/backup_processor.go

```diff
@@ -193,280 +201,417 @@ func runBackupProcessor(
-// KV level. It also enables tenant access to `userfile` destinations, which store the
-// data in SQL. However, this does come at a cost since we need to incur an extra copy
-// of the data and therefore increase network and memory utilization in the cluster.
-func writeFile(
-	ctx context.Context,
-	file roachpb.ExportResponse_File,
-	defaultStore cloud.ExternalStorage,
-	storeByLocalityKV map[string]cloud.ExternalStorage,
-	enc *roachpb.FileEncryptionOptions,
-) error {
-	if defaultStore == nil {
-		return errors.New("no default store created when writing SST")
-	}
+type sstSinkConf struct {
+	progCh         chan execinfrapb.RemoteProducerMetadata_BulkProcessorProgress
+	targetFileSize int64
+	enc            *roachpb.FileEncryptionOptions
+	id             base.SQLInstanceID
+}
 
-	data := file.SST
-	locality := file.LocalityKV
+type sstSink struct {
+	dest             cloud.ExternalStorage
+	conf             sstSinkConf
+	sstWriter        *storage.SSTWriter
+	bufferedSST      storage.MemFile
+	bufferedFiles    []BackupManifest_File
+	bufferedRevStart hlc.Timestamp
+	completedSpans   int32
+}
 
-	exportStore := defaultStore
-	if localitySpecificStore, ok := storeByLocalityKV[locality]; ok {
-		exportStore = localitySpecificStore
+func (s *sstSink) flush(ctx context.Context) error {
+	if s.sstWriter == nil {
+		return nil
+	}
+	if err := s.sstWriter.Finish(); err != nil {
+		return err
 	}
 
-	if enc != nil {
+	data := s.bufferedSST.Bytes()
+	if s.conf.enc != nil {
 		var err error
-		data, err = storageccl.EncryptFile(data, enc.Key)
+		data, err = storageccl.EncryptFile(data, s.conf.enc.Key)
 		if err != nil {
 			return err
 		}
 	}
 
-	if err := cloud.WriteFile(ctx, exportStore, file.Path, bytes.NewReader(data)); err != nil {
+	name := fmt.Sprintf("%d.sst", builtins.GenerateUniqueInt(s.conf.id))
+	if err := cloud.WriteFile(ctx, s.dest, name, bytes.NewReader(data)); err != nil {
 		log.VEventf(ctx, 1, "failed to put file: %+v", err)
 		return errors.Wrap(err, "writing SST")
 	}
+
+	for i := range s.bufferedFiles {
+		s.bufferedFiles[i].Path = name
+	}
+
+	progDetails := BackupManifest_Progress{
+		RevStartTime:   s.bufferedRevStart,
+		Files:          s.bufferedFiles,
+		CompletedSpans: s.completedSpans,
+	}
+	var prog execinfrapb.RemoteProducerMetadata_BulkProcessorProgress
+	details, err := gogotypes.MarshalAny(&progDetails)
+	if err != nil {
+		return err
+	}
+	prog.ProgressDetails = *details
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case s.conf.progCh <- prog:
+	}
+
+	s.sstWriter = nil
+	s.bufferedSST.Reset()
+	s.bufferedFiles = nil
+	s.bufferedRevStart.Reset()
+	s.completedSpans = 0
+
+	return nil
+}
+
+func (s *sstSink) write(ctx context.Context, resp returnedSST) error {
+	span := resp.f.Span
+
+	// If this span starts before the last buffered span ended, we need to flush
+	// since it overlaps but SSTWriter demands writes in-order.
+	// TODO(dt): consider buffering resp until _next_ `write` to allow minimal
```

Comment:
It seems like the concurrency in the producer could make us miss many merging opportunities in the sstSink. This is also a problem that RESTORE faces with the introduced concurrency. I think this is different than what this TODO is saying; but I may have misunderstood.

I wonder if we could give the channel a bit more buffer to give a bit of a lead to the producer, then the sstWriter could pull off the SSTs and buffer them in a sorted data structure. This would certainly increase the memory overhead of this buffering, but I worry that we won't see much improvement in the merging otherwise.

Alternatively, maybe we can sort the buffered files by span in the sstSink during flush (they should all be non-overlapping) and merge them there rather than in write?

Reply:
That is what this TODO is about: right now if a later file gets to the sink before an earlier one, the sink is currently forced to flush the later, then add the earlier to its buffer. The todo is suggesting the sink buffer the received file before pushing its contents into the SST writer, so that if the next (or next n) files show up and are indeed earlier, we can flush them first instead of being forced to start a new file.

I just didn't want to do that in this change since it looks a little bit different after the direct-write-ssts change that is queued up next.